### ISSUE TITLE
PFCG roles (authorizations)

### ISSUE THREAD

**Issue by larshp** [larshp/abapGit#827]
Currently PFCG roles are not supported, as they are not recorded in TADIR

However we should find some way to make it work

**Comment by EduardoCopat** [comment]
Some questions:

- Isn't PFCG roles/authorizations a customizing activity? I thought abapGit should only parse `SUSO`.
- How are we linking a package to them?

`AGR_*` tables is where roles are stored.

**Comment by larshp** [comment]
yeah, thats the issue, not sure how to handle it, mainly as the roles does not belong to a package

however I think this is a feature we need in abapGit

**Comment by EduardoCopat** [comment]
Ideas: Since this is not connected to any package, would we have an advanced option to export and import to zip? Similar to the transport feature?

**Comment by larshp** [comment]
One approach could be somehow linking the role to a package, eg. in a custom screen in ZABAPGIT, and then add this in .abapgit.xml

**Comment by BUPA** [comment]
I just found this issue as im searched for a possibility to add pfcg roles to a repository.

There's an function module 'PRGN_DOWNLOAD_AGRS' which in fact already is able to transfer a role into a filestream.
this is used by transaction PFCG's download function. 
To reupload it into the SAP System there's module 'PRGN_EXECUTE_UPLOAD' which can import the file created by the first module back into the SAP system.

Unfortunately I'm new with abapgit and not sure how I would be able to integrate this correctly into the repo.

**Comment by larshp** [comment]
abapGit starts by looking at objects via TADIR, and PFCG roles are not recorded in TADIR.

This might be possible to solve by including the role in a business configuration set, https://github.com/larshp/abapGit/issues/1004 as these are added in TADIR

**Comment by larshp** [comment]
It looks like the creation of the BC set from a customizing transport is quite simple,

![image](https://user-images.githubusercontent.com/5888506/41047676-667c7f06-69ad-11e8-93ac-e5014d7718cd.png)

it might be a bit more work to implement, but it might fix a lot with a bit larger investment

**Comment by BUPA** [comment]
Sorry but maybe I'm just on the wrong path. I'm considering the workflow how to use it in the first place.

The BC-Set would have to be assigned to an proper package (like the rest of the application) to be caught by TADIR access. Therefore it would have to be a transportable object. 
As soon as the transport would be released (like it would be when leaving development) the BC set is fixed. It's content is frozen to the moment of the export from DEV system. 

Therefore the handling of roles would be 
1) create a transport (ok this one you need for every change) 
2) create a new BC Set of it
3) Commit and Push
As soon as someone changing an existing role without the BC Set creation, it would be missed by abapgit as the initial BC set is already frozen.

Tbh the worklfow isn't this uncomfortable as i thought in the first place. And the BCSet point would be replaced by a "add to repository" step. Only existing roles could be handled easier.

**Comment by BUPA** [comment]
If this would mean such big changes to heart of git im with you. Sorry for all the discussions.

**Comment by larshp** [comment]
discussions/ideas/suggestions are good

If somebody implements the suggestion above, I have no problems merging it. But most likely it will be deprecated when/if the BC-Sets are implemented

**Comment by jrodriguez-rc** [comment]
I was trying to include a role in a BC Set, but it doesn't allow roles
![image](https://user-images.githubusercontent.com/56556686/79332550-4ebcda00-7f1d-11ea-9cb6-ea8b2ff80bc3.png)

**Comment by gardian12** [comment]
Hello together, I need to resume the initial feature. How to handle roles?
Here are the facts:
- Roles can't be handled with BC Sets
- Roles are customizing and not linked to a development package
- Roles are essential part of an application and should be delivered as well
- Roles can be easily handled as files, see transaction PFCG (upload / download function)

Therefore it would be highly recommended also to cover PFCG roles by abapGit
Without disturbing the core of abapGit it would be possible.

As an example: Customizing files, like roles, could be stored with a separated folder, parallel to /src/.
Manual selection of role to be added to the repo would be needed any way.

Hope you can reopen, the feature request with this insights.
